### PR TITLE
Week four slides fixes

### PR DESCRIPTION
**Pull request by JasonGibson274**
This is by no means done I just want someone else to verify that the flow of the slides is correct.

### CODE REVIEW

File: slides/week4/week4.org [hunk 1]
```diff
@@ -74,10 +237,27 @@ A* new_var = (A*)old_var;
...
+#+BEGIN_NOTES
+OpenSSL cryptography library bug. no bounds checking allowed for buffer overflow.
+#+END_NOTES
+- Heartbleed
+- https://xkcd.com/1354/
+
+
+* Memory Leak
+#+BEGIN_NOTES
+run a program that creates a lot of doubles while showing your memory usage. files may corrupt so make sure everything is closed beforehand
```
Comment (jgkamat) [inline slides/week4/week4.org]: If you turn off swap or let swap fill, linux will actually kill a process to save the system (oom killer). If this happens, you can show the dmesg output to show the log of the killed process.

**Comment by JasonGibson274** [comment]
Let me know what everyone thinks

File: slides/week4/week4.org [hunk 2]
```diff
@@ -3,57 +3,336 @@
...
  - *The heap:* your space to do whatever you want
-  - You can ask for memory from the OS and return it at runtime
+  - You can ask for memory from the OS and return it
+    - The heap is managed at runtime
   - You have to keep track of what you've allocated to avoid memory leaks
 - At an even higher level, all the memory in your computer is a huge, indexable, iterable space, divided into bytes
 
+** What is a byte
+#+BEGIN_NOTES
+They may never have seen binary before. 26, 17 do not talk about how binary works
```
Comment (jgkamat) [inline slides/week4/week4.org]: > 26, 17 

huh?

File: slides/week4/week4.org [hunk 3]
```diff
@@ -3,57 +3,336 @@
...
+  shared_ptr<object> s_ptr1 = make_shared<object>();
+  {
+    shared_ptr<object> s_ptr2 = make_shared<object>();
+  }
+  shared_ptr<object> s_ptr3 = make_shared<object>();
+  // <--- HERE
+}
+#+END_SRC
+#+ATTR_HTML: :width 70%
+[[file:https://i.imgur.com/qLddS8b.png]]
```
Comment (jgkamat) [inline slides/week4/week4.org]: This image is misleading. `s_ptr2` get's deallocated from the stack (popped off), so `s_ptr3` would take it's place, not be placed after the `???`

**Review by jgkamat — CHANGES_REQUESTED** [review]
Overall, looking very good! :smile: 

Got a few nitpicks though.

File: slides/week4/week4.org (hunk 2 above)
Comment (JasonGibson274) [reply slides/week4/week4.org]: What those binary numbers are. I made the comment bettter

File: slides/week4/week4.org (hunk 3 above)
Comment (JasonGibson274) [reply slides/week4/week4.org]: valid point. Changed graphic

**Review by jgkamat — DISMISSED** [review]
LGTM! :)

File: slides/week4/week4.org [hunk 4]
```diff
@@ -3,57 +3,336 @@
...
+
+** Object lifetimes stack
+#+BEGIN_NOTES
+Picture on next slide. talk about how in the below example the int is allocated on the stack. mention what stack overflow is
+#+END_NOTES
+- All variables have different lifetimes depending on where they are allocated in memory
+- Every method call has its own copy of the variables on the stack
+#+BEGIN_SRC c++
+#include <iostream>
+void increment(int a) {
```
Comment (chachmu) [inline slides/week4/week4.org:58]: I may just be dumb but I don't see a difference between this code and the example that starts at line 73

Comment (JasonGibson274) [reply slides/week4/week4.org:58]: It is exactly the same but the picture did not fit so I pushed it to the next slide and copied the code. It happens a lot.

**Comment by JasonGibson274** [comment]
@jgkamat can I get an approval for merge?

**Review by rmkeezer — APPROVED** [review]
LGTM

**Review by jgkamat — APPROVED** [review]
Already approved before, but LGTM again

**Pull request merged**